File: app/src/main/java/com/example/jetpackdemo/data/model/InteractiveModels.kt
package com.example.jetpackdemo.data.model

import com.google.gson.annotations.SerializedName

// === Interactive Session (GET /api/interactive/:subtopicId or /course/:courseId/next) ===
data class InteractiveSessionResponse(
    val subtopic: InteractiveSubtopic? = null,
    val questions: List<InteractiveQuestion> = emptyList(),
    @SerializedName("hearts_remaining")
    val heartsRemaining: Int = 3,
    val attempts: Int = 0,
    @SerializedName("is_completed")
    val isCompleted: Boolean = false,
    @SerializedName("course_completed")
    val courseCompleted: Boolean = false,
    val message: String? = null
)

data class InteractiveSubtopic(
    val id: String,
    val title: String,
    val content: Any? = null,
    @SerializedName("course_id")
    val courseId: String? = null
)

data class InteractiveQuestion(
    val id: String,
    @SerializedName("question_text")
    val questionText: String,
    val options: List<String> = emptyList(),
    val hint: String? = null,
    val type: String = "mcq"
)

// === Verify Answer (POST /api/interactive/:subtopicId/verify) ===
data class VerifyAnswerRequest(
    val questionId: String,
    val answer: String
)

data class VerifyAnswerResponse(
    val correct: Boolean,
    @SerializedName("correct_answer")
    val correctAnswer: String? = null,
    val hint: String? = null,
    @SerializedName("hearts_remaining")
    val heartsRemaining: Int = 0,
    val result: String? = null,
    @SerializedName("game_over")
    val gameOver: Boolean = false,
    @SerializedName("is_subtopic_completed")
    val isSubtopicCompleted: Boolean = false
)

// === Chat (POST /api/interactive/:subtopicId/chat) ===
data class InteractiveChatRequest(
    val message: String,
    val provider: String = "Groq"
)

data class InteractiveChatResponse(
    @SerializedName("ai_response")
    val aiResponse: String
)

// === Course Chat (POST /api/interactive/course/:courseId/chat) ===
data class CourseChatRequest(
    val message: String,
    val provider: String = "Groq",
    @SerializedName("session_id")
    val sessionId: String? = null
)

data class CourseChatResponse(
    @SerializedName("session_id")
    val sessionId: String,
    @SerializedName("ai_response")
    val aiResponse: String,
    val provider: String? = null
)

// === Course Practice (POST /api/interactive/course/:courseId/practice) ===
data class CoursePracticeRequest(
    val focus: String = "general revision",
    val provider: String = "Groq"
)

data class CoursePracticeResponse(
    @SerializedName("course_id")
    val courseId: String,
    val focus: String,
    val provider: String,
    val questions: List<PracticeQuestion> = emptyList()
)

data class PracticeQuestion(
    val question: String,
    val answer: String,
    val explanation: String = "",
    val type: String = "concept"
)

// === Audio (GET /api/audio/:subtopicId, /api/audio/course/:courseId) ===
data class GetAudioResponse(
    val audio: AudioData,
    val generated: Boolean
)

data class AudioData(
    val id: String? = null,
    val type: String = "subtopic",
    @SerializedName("subtopic_id")
    val subtopicId: String? = null,
    @SerializedName("subtopic_title")
    val subtopicTitle: String? = null,
    @SerializedName("unit_title")
    val unitTitle: String? = null,
    @SerializedName("course_title")
    val courseTitle: String? = null,
    @SerializedName("course_id")
    val courseId: String? = null,
    val script: String = "",
    @SerializedName("segment_count")
    val segmentCount: Int = 0,
    @SerializedName("estimated_duration")
    val estimatedDuration: Int = 0,
    @SerializedName("tts_provider")
    val ttsProvider: String? = null,
    val voice: String? = null,
    @SerializedName("audio_url")
    val audioUrl: String = "",
    @SerializedName("imagekit_file_id")
    val imagekitFileId: String? = null,
    @SerializedName("file_size_bytes")
    val fileSizeBytes: Long = 0,
    @SerializedName("generated_at")
    val generatedAt: String? = null
)

// === Chat message for local UI state ===
data class ChatMessage(
    val text: String,
    val isUser: Boolean,
    val timestamp: Long = System.currentTimeMillis()
)

// =====================================================================
// === Content-First Interactive Flow Models ============================
// =====================================================================

// GET /api/interactive/course/:courseId/next-content
data class ContentResponse(
    @SerializedName("course_completed")
    val courseCompleted: Boolean = false,
    val subtopic: InteractiveSubtopic? = null,
    @SerializedName("questions_ready")
    val questionsReady: Boolean = false,
    @SerializedName("hearts_remaining")
    val heartsRemaining: Int = 3,
    val attempts: Int = 0,
    @SerializedName("is_completed")
    val isCompleted: Boolean = false,
    @SerializedName("course_progress")
    val courseProgress: CourseProgress? = null,
    val message: String? = null
)

data class CourseProgress(
    val completed: Int = 0,
    val total: Int = 0
)

// GET /api/interactive/:subtopicId/quiz
data class QuizResponse(
    @SerializedName("subtopic_id")
    val subtopicId: String,
    val questions: List<QuizQuestion> = emptyList(),
    @SerializedName("hearts_remaining")
    val heartsRemaining: Int = 3,
    @SerializedName("is_completed")
    val isCompleted: Boolean = false
)

data class QuizQuestion(
    val id: String,
    @SerializedName("question_text")
    val questionText: String,
    @SerializedName("question_type")
    val questionType: String = "mcq",
    val options: List<String> = emptyList(),
    @SerializedName("correct_answer")
    val correctAnswer: String,
    val hint: String = "",
    val explanation: String = "",
    val position: Int = 0
)

// POST /api/interactive/:subtopicId/submit-quiz
data class SubmitQuizRequest(
    val answers: List<QuizAnswerItem>
)

data class QuizAnswerItem(
    @SerializedName("question_id")
    val questionId: String,
    @SerializedName("user_answer")
    val userAnswer: String,
    @SerializedName("is_correct")
    val isCorrect: Boolean
)

data class SubmitQuizResponse(
    @SerializedName("subtopic_id")
    val subtopicId: String,
    @SerializedName("total_questions")
    val totalQuestions: Int,
    @SerializedName("correct_count")
    val correctCount: Int,
    @SerializedName("wrong_count")
    val wrongCount: Int,
    @SerializedName("hearts_remaining")
    val heartsRemaining: Int,
    @SerializedName("is_completed")
    val isCompleted: Boolean,
    @SerializedName("is_perfect")
    val isPerfect: Boolean,
    @SerializedName("xp_earned")
    val xpEarned: Int
)

// === Hub Activity Models ===

// POST /api/interactive/hub/log
data class HubActivityRequest(
    @SerializedName("course_id")
    val courseId: String,
    @SerializedName("subtopic_id")
    val subtopicId: String? = null,
    @SerializedName("feature_type")
    val featureType: String,
    val title: String = ""
)

// GET /api/interactive/hub/history/:courseId
data class HubHistoryResponse(
    @SerializedName("course_id")
    val courseId: String,
    val activities: List<HubActivity> = emptyList(),
    @SerializedName("subtopic_statuses")
    val subtopicStatuses: Map<String, SubtopicFeatureStatus> = emptyMap(),
    @SerializedName("generated_items")
    val generatedItems: List<GeneratedItem> = emptyList()
)

data class HubActivity(
    val id: String,
    @SerializedName("user_id")
    val userId: String,
    @SerializedName("course_id")
    val courseId: String,
    @SerializedName("subtopic_id")
    val subtopicId: String? = null,
    @SerializedName("feature_type")
    val featureType: String,
    val title: String = "",
    @SerializedName("created_at")
    val createdAt: String? = null
)

data class SubtopicFeatureStatus(
    val title: String = "",
    @SerializedName("unit_title")
    val unitTitle: String? = null,
    @SerializedName("has_flashcards")
    val hasFlashcards: Boolean = false,
    @SerializedName("has_notes")
    val hasNotes: Boolean = false,
    @SerializedName("has_audio")
    val hasAudio: Boolean = false,
    @SerializedName("has_content")
    val hasContent: Boolean = false
)

data class GeneratedItem(
    val type: String = "",
    @SerializedName("subtopic_id")
    val subtopicId: String? = null,
    @SerializedName("subtopic_title")
    val subtopicTitle: String = "",
    @SerializedName("unit_title")
    val unitTitle: String? = null,
    @SerializedName("course_id")
    val courseId: String? = null,
    @SerializedName("generated_at")
    val generatedAt: String? = null,
    @SerializedName("audio_url")
    val audioUrl: String? = null,
    @SerializedName("estimated_duration")
    val estimatedDuration: Int = 0,
    @SerializedName("card_count")
    val cardCount: Int = 0
)

// === Local quiz state for client-side verification ===
data class QuizAttemptResult(
    val questionId: String,
    val questionText: String,
    val userAnswer: String,
    val correctAnswer: String,
    val isCorrect: Boolean,
    val explanation: String = "",
    val hint: String = ""
)

File: ailearner_flutter/lib/features/interactive/data/interactive_models.dart
class InteractiveQuestion {
  const InteractiveQuestion({
    required this.id,
    required this.questionText,
    required this.type,
    this.options = const [],
    this.hint,
  });

  final String id;
  final String questionText;
  final String type;
  final List<String> options;
  final String? hint;

  factory InteractiveQuestion.fromJson(Map<String, dynamic> json) {
    final rawOptions = (json['options'] as List?) ?? <dynamic>[];
    return InteractiveQuestion(
      id: json['id'] as String? ?? '',
      questionText: json['question_text'] as String? ?? '',
      type: (json['type'] as String? ?? 'mcq').toLowerCase(),
      options: rawOptions.map((e) => e.toString()).toList(),
      hint: json['hint'] as String?,
    );
  }
}

class InteractiveSubtopic {
  const InteractiveSubtopic({
    required this.id,
    required this.title,
    required this.content,
    required this.courseId,
  });

  final String id;
  final String title;
  final String content;
  final String courseId;

  factory InteractiveSubtopic.fromJson(Map<String, dynamic> json) {
    return InteractiveSubtopic(
      id: json['id'] as String? ?? '',
      title: json['title'] as String? ?? '',
      content: json['content'] as String? ?? '',
      courseId: json['course_id'] as String? ?? '',
    );
  }
}

class InteractiveSession {
  const InteractiveSession({
    required this.subtopic,
    required this.questions,
    required this.heartsRemaining,
    required this.attempts,
    required this.isCompleted,
    required this.courseCompleted,
    this.message,
  });

  final InteractiveSubtopic? subtopic;
  final List<InteractiveQuestion> questions;
  final int heartsRemaining;
  final int attempts;
  final bool isCompleted;
  final bool courseCompleted;
  final String? message;

  factory InteractiveSession.fromJson(Map<String, dynamic> json) {
    final rawQuestions = (json['questions'] as List?) ?? <dynamic>[];
    return InteractiveSession(
      subtopic: json['subtopic'] is Map<String, dynamic>
          ? InteractiveSubtopic.fromJson(json['subtopic'] as Map<String, dynamic>)
          : null,
      questions: rawQuestions
          .whereType<Map>()
          .map((e) => InteractiveQuestion.fromJson(e.cast<String, dynamic>()))
          .toList(),
      heartsRemaining: (json['hearts_remaining'] as num?)?.toInt() ?? 3,
      attempts: (json['attempts'] as num?)?.toInt() ?? 0,
      isCompleted: json['is_completed'] as bool? ?? false,
      courseCompleted: json['course_completed'] as bool? ?? false,
      message: json['message'] as String?,
    );
  }
}

class VerifyAnswerResult {
  const VerifyAnswerResult({
    required this.correct,
    this.correctAnswer,
    this.hint,
    required this.heartsRemaining,
    required this.gameOver,
    required this.isSubtopicCompleted,
    this.result,
  });

  final bool correct;
  final String? correctAnswer;
  final String? hint;
  final int heartsRemaining;
  final bool gameOver;
  final bool isSubtopicCompleted;
  final String? result;

  factory VerifyAnswerResult.fromJson(Map<String, dynamic> json) {
    return VerifyAnswerResult(
      correct: json['correct'] as bool? ?? false,
      correctAnswer: json['correct_answer'] as String?,
      hint: json['hint'] as String?,
      heartsRemaining: (json['hearts_remaining'] as num?)?.toInt() ?? 0,
      gameOver: json['game_over'] as bool? ?? false,
      isSubtopicCompleted: json['is_subtopic_completed'] as bool? ?? false,
      result: json['result'] as String?,
    );
  }
}

class GeneratedNotes {
  const GeneratedNotes({
    required this.summary,
    required this.keyPoints,
    required this.raw,
  });

  final String summary;
  final List<String> keyPoints;
  final Map<String, dynamic> raw;

  factory GeneratedNotes.fromJson(Map<String, dynamic> json) {
    final notes = json['notes'] is Map<String, dynamic>
        ? json['notes'] as Map<String, dynamic>
        : json;
    final points = (notes['key_points'] as List?)?.map((e) => e.toString()).toList() ?? <String>[];
    return GeneratedNotes(
      summary: notes['summary'] as String? ?? '',
      keyPoints: points,
      raw: notes,
    );
  }
}

class FlashcardItem {
  const FlashcardItem({
    required this.id,
    required this.front,
    required this.back,
    required this.cardType,
    required this.position,
  });

  final String id;
  final String front;
  final String back;
  final String cardType;
  final int position;

  factory FlashcardItem.fromJson(Map<String, dynamic> json) {
    return FlashcardItem(
      id: json['id'] as String? ?? '',
      front: json['front'] as String? ?? '',
      back: json['back'] as String? ?? '',
      cardType: json['card_type'] as String? ?? 'concept',
      position: (json['position'] as num?)?.toInt() ?? 0,
    );
  }
}

class ChatMessage {
  const ChatMessage({
    required this.text,
    required this.isUser,
  });

  final String text;
  final bool isUser;
}
